### PR TITLE
Common code supporting HMEM

### PR DESCRIPTION
**Pull request by iziemba**
Needed for https://github.com/ofiwg/libfabric/pull/5885.

### CODE REVIEW

**Comment by iziemba** [comment]
Regarding the comment about AMD support I put in https://github.com/ofiwg/libfabric/pull/5885, I have opened a draft PR for how an AMD ROCr memory monitor can be implemented.

File: Makefile.am [hunk 1]
```diff
@@ -39,6 +40,7 @@ rdmainclude_HEADERS =
 
 # internal utility functions shared by in-tree providers:
 common_srcs =				\
+	src/hmem.c			\
```
Comment (iziemba) [inline Makefile.am:43]: I believe I need to also add "include/ofi_hmem.h" to the src_libfabric_la_SOURCES section.

File: Makefile.am [hunk 2]
```diff
@@ -70,6 +70,7 @@ common_srcs =				\
 	prov/util/src/util_mem_monitor.c\
 	prov/util/src/util_mem_hooks.c	\
 	prov/util/src/util_mr_cache.c	\
+	prov/util/src/util_cuda_mem_monitor.c \
```
Comment (iziemba) [inline Makefile.am:73]: Need to update libfabric.vcxproj

File: prov/util/src/util_mr_cache.c [hunk 3]
```diff
@@ -310,6 +311,17 @@ int ofi_mr_cache_search(struct ofi_mr_cache *cache, const struct fi_mr_attr *att
 
 		cache->search_cnt++;
 		*entry = cache->storage.find(&cache->storage, &info);
+
+		if (*entry &&
+		    !ofi_monitor_valid(cache->monitor,
+				       (const void *)(*entry)->info.iov.iov_base,
+				       (*entry)->info.iov.iov_len)) {
+			ofi_monitor_unsubscribe(cache->monitor,
```
Comment (rajachan) [inline prov/util/src/util_mr_cache.c:319]: Do you need this unsubscribe? The entry is going to get purged if it is not usable anyway.

File: src/hmem.c [hunk 4]
```diff
@@ -184,3 +184,54 @@ ssize_t ofi_copy_to_hmem_iov(const struct iovec *hmem_iov,
 				     hmem_iov_offset, src, size,
 				     OFI_COPY_BUF_TO_IOV);
 }
+
+int ofi_get_hmem_iface(const void *buf, enum fi_hmem_iface *iface)
```
Comment (rajachan) [inline src/hmem.c:188]: I ask this without reading through the corresponding verbs/rxm PR, but where do you expect to use this? If an application is already giving this information about a buffer, should that not be all a provider needs? If an application is passing in a non-system buffer without setting the right attribute, that will be an erroneous use of the API.

**Comment by rajachan** [comment]
This PR will also make https://github.com/ofiwg/libfabric/pull/5808 redundant, so I can close it out once this merges. Based on the discussion in https://github.com/ofiwg/libfabric/issues/5789, we'd want to make the memhooks the primary mechanism and the buffer ID check you have here the fallback.

File: prov/util/src/util_mr_cache.c [hunk 5]
```diff
@@ -296,6 +296,7 @@ int ofi_mr_cache_search(struct ofi_mr_cache *cache, const struct fi_mr_attr *att
 	       attr->mr_iov->iov_base, attr->mr_iov->iov_len);
 
 	info.iov = *attr->mr_iov;
+	info.iface = attr->iface;
```
[the remark this answers is not in the document]
Comment (rajachan) [reply prov/util/src/util_mr_cache.c:300]: Ignore this comment, this is the right place.

File: Makefile.am (hunk 1 above)
Comment (aingerson) [inline Makefile.am:43]: What are your thoughts on keeping device-specific ops separate (ie keep all the cuda ops in one file)? Just thinking that if we add more ops (open/close/get IPC handle, alloc/free, etc) and add more devices, this file might start to get a bit confusing.

File: src/hmem.c [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+}
+
+static int ofi_hmem_cuda_memcpy_dtoh(void *host, const void *dev, size_t size)
+{
+	return -FI_ENOSYS;
+}
+
+#endif /* HAVE_LIBCUDA */
+
+#define FI_HMEM_IFACE_TYPE_COUNT 2
```
[the remark this answers is not in the document]
Comment (iziemba) [reply src/hmem.c:102]: Sounds good. I'll add in a FI_HMEM_MAX value.

[the remark this answers is not in the document]
Comment (shefty) [reply src/hmem.c:102]: No - don't.  We can't put a value in the public header and allow it to change between libfabric versions.

File: src/hmem.c [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+	return hmem_ops[iface].memcpy_htod(dev, host, size);
+}
+
+static inline int ofi_hmem_memcpy_dtoh(void *host, const void *dev,
+				       size_t size, enum fi_hmem_iface iface)
+{
+	return hmem_ops[iface].memcpy_dtoh(host, dev, size);
+}
+
+static ssize_t ofi_copy_hmem_iov_buf(const struct iovec *hmem_iov,
```
[the remark this answers is not in the document]
Comment (iziemba) [reply src/hmem.c:127]: Correct.

File: prov/util/src/util_mr_cache.c (hunk 3 above)
Comment (iziemba) [reply prov/util/src/util_mr_cache.c:319]: The unsubscribe will purge the entry from the CUDA memory monitor. Since the CUDA memory monitor does not actively check regions, it needs the MR cache to purge entries. Or, entries will be purged if a threshold is exceeded.

File: src/hmem.c (hunk 4 above)
Comment (iziemba) [reply src/hmem.c:188]: I have the requirement of supporting rxm;verbs where the user does not register device memory for local operations. Since an MR is no longer being provided for operations, ofi_get_hmem_iface() is used to rediscover the buffer type.

Note that this capability is being advertised to users by having the FI_HMEM capability set and the FI_MR_HMEM MR mode bit unset.

File: src/hmem.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (iziemba) [reply src/hmem.c:102]: Sounds good.

File: prov/util/src/util_mem_monitor.c [hunk 8]
```diff
@@ -170,6 +171,16 @@ void ofi_monitor_del_cache(struct ofi_mr_cache *cache)
 	pthread_mutex_unlock(&monitor->lock);
 }
 
+/* Must be called holding monitor lock */
+bool ofi_monitor_valid(struct ofi_mem_monitor *monitor, const void *addr,
+		       size_t len)
+{
+	if (!monitor->valid)
+		return true;
```
Comment (shefty) [inline prov/util/src/util_mem_monitor.c:187]: If a valid callback is not set by the monitor, we should set the monitor->valid() callback to a function that always returns true, rather than adding this check.  This wrapper itself isn't really needed then.

File: prov/util/src/util_mr_cache.c (hunk 3 above)
Comment (shefty) [reply prov/util/src/util_mr_cache.c:319]: There may be multiple entries in the cache that are covered by the new region, all of which need to be removed in order that future searches will work.  Try to match the behavior of the original code -- check first for a usable entry and goto hit in that case.  If the entry is not usable, we need the while loop below to purge all unusable entries.

File: Makefile.am (hunk 1 above)
Comment (iziemba) [reply Makefile.am:43]: I could do something like:
- CUDA: src/cuda_hmem.c with ofi_cuda_hmem.h
- ROCr (AMD future work): src/rocr_hmem.c with ofi_rocr_hmem.h

Mainly, as long as I don't lose ofi_copy_from_hmem_iov() and ofi_copy_to_hmem_iov(), I am ok with any change.

**Comment by iziemba** [comment]
> This PR will also make #5808 redundant, so I can close it out once this merges. Based on the discussion in #5789, we'd want to make the memhooks the primary mechanism and the buffer ID check you have here the fallback.

The one major value add https://github.com/ofiwg/libfabric/pull/5808 has over this CUDA memory monitor cache implement is the ability to have a single MR cache serve both host and CUDA buffers. For example, in https://github.com/ofiwg/libfabric/pull/5885, I had to have an MR cache per HMEM type. I think having a cache which serves both host and device memory makes sense. So either https://github.com/ofiwg/libfabric/pull/5808 should be used, or maybe refactor the MR cache to be able to subscribe to multiple memory monitors. Whatever the end solution is, it should also easily accommodate AMD GPUs. I have opened up a draft PR where I implemented a memory monitor for AMD GPUs: https://github.com/ofiwg/libfabric/pull/5887.

Thoughts?

File: src/hmem.c [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+		}
+
+		hmem_buf = (char *)hmem_iov[i].iov_base + hmem_iov_offset;
+		len -= hmem_iov_offset;
+
+		len = MIN(len, size);
+		if (dir == OFI_COPY_BUF_TO_IOV)
+			ret = ofi_hmem_memcpy_htod(hmem_buf, (char *)buf + done,
+						   len, hmem_iface);
+		else if (dir == OFI_COPY_IOV_TO_BUF)
```
[the remark this answers is not in the document]
Comment (iziemba) [reply src/hmem.c:153]: Will fix.

File: prov/util/src/util_mem_monitor.c (hunk 8 above)
Comment (iziemba) [reply prov/util/src/util_mem_monitor.c:187]: Sounds good.

File: prov/util/src/util_mr_cache.c (hunk 3 above)
Comment (iziemba) [reply prov/util/src/util_mr_cache.c:319]: I'll work on cleaning this up. One thing that may not be clear at first glance is that the CUDA memory monitor monitors an entire user allocated region. When ofi_monitor_unsubscribe() is called, the CUDA memory monitor will free the corresponding monitored region and call ofi_monitor_notify() for the entire region at the end of the free (ofi_cuda_mm_entry_free()). I believe this should cleanup all stale entries in all subscribed MR caches. But, this back and forth between the MR cache and CUDA memory monitor may not be ideal.

Comment (shefty) [reply prov/util/src/util_mr_cache.c:319]: The entries that a new region may match may be a mix of device and host buffers.

Comment (iziemba) [reply prov/util/src/util_mr_cache.c:319]: Will fix.

File: Makefile.am (hunk 1 above)
Comment (aingerson) [reply Makefile.am:43]: I would prefer separation if that's ok with you and @shefty 
I have been told I'm bad at naming things so I will let you or him figure out the best names for the files :P

File: prov/util/src/util_cuda_mem_monitor.c [hunk 10]
```diff
@@ -0,0 +1,416 @@
...
+
+#include <cuda.h>
+#include <cuda_runtime.h>
+
+struct ofi_cuda_mm_entry {
+	struct iovec iov;
+	uint64_t id;
+	struct dlist_entry lru_entry;
+	struct ofi_rbnode *node;
+};
```
Comment (shefty) [inline prov/util/src/util_cuda_mem_monitor.c:48]: This looks like it's duplicating the functionality of the mr cache.  Specifically, this is similar to ofi_mr_entry.  Only the cache should need to maintain a rbtree and lru list.  We shouldn't require 2 searches for the same address range.  If hmem requires storing some additional amount of data per ofi_mr_entry (such as an id), let's put that data there.

The ofi_mr_entry is currently broken into 2 pieces.  There's the portion used by the cache itself, and a secondary data section for use by the provider.  It seems that we just need a third section that's HMEM specific.  Although, we could allocate this dynamically, it doesn't seem worth it at this point.  For example:

```
union ofi_mr_hmem_info {
    uint64_t cuda_id;
};

struct ofi_mr_entry {
    struct ofi_mr_info info;
    void  *storage_context;
    unsigned int subscribed:1;
    int use_cnt;
    struct dlist_entry list_entry;
    enum fi_hmem_iface iface;
    union ofi_mr_hmem_info hmem_info;
    uint8_t data[];
};
```

This moves the iface from ofi_mr_info here.  I haven't looked close enough at the patches to see if ofi_mr_info really needs this data, rather than having it here.  This requires changing the monitor valid() handler to accept the hmem_info as input.  Monitors that do not use this data can ignore it.

File: prov/util/src/util_mem_monitor.c [hunk 11]
```diff
@@ -140,6 +144,8 @@ int ofi_monitor_add_cache(struct ofi_mem_monitor *monitor,
 			ret = ofi_uffd_init();
 		else if (monitor == memhooks_monitor)
 			ret = ofi_memhooks_init();
+		else if (monitor == cuda_monitor)
+			ret = FI_SUCCESS;
```
Comment (shefty) [inline prov/util/src/util_mem_monitor.c:148]: I would modify the cache to support 2 (or more) monitors, up to 1 per fi_hmem_iface.  At a minimum, I would allow the host memory monitor, plus one device monitor.  Though it's likely barely more work or overhead to handle all monitors at once.  The monitor would be selected based on the ofi_mr_entry hmem_iface value for each region.

We need a single cache for each virtual/device address range that we're dealing with.  If we have a unified virtual address range, then those addresses should all be stored in the same cache.  If we're dealing with GPU specific addresses (where the value returned could overlap with a value returned through a host memory call), then those need to go into a separate cache.

If we can defer dealing with multiple caches until a later patch set, that would be preferable to me.  But I think that information needs to come down from the application through the API, with the providers selecting the correct cache.

File: prov/util/src/util_cuda_mem_monitor.c (hunk 10 above)
Comment (iziemba) [reply prov/util/src/util_cuda_mem_monitor.c:48]: Sounds good.

File: prov/util/src/util_mem_monitor.c (hunk 11 above)
Comment (iziemba) [reply prov/util/src/util_mem_monitor.c:148]: I was starting to look into this change. Would you like the "cache to support 2 (or more) monitors, up to 1 per fi_hmem_iface" change as a part of this patch set or a future PR?

File: libfabric.vcxproj [hunk 12]
```diff
@@ -568,6 +568,7 @@
     <ClCompile Include="prov\util\src\util_mem_monitor.c" />
     <ClCompile Include="prov\util\src\util_mem_hooks.c" />
     <ClCompile Include="prov\util\src\util_mr_cache.c" />
+    <ClCompile Include="prov\util\src\util_cuda_mem_monitor.c" />
```
Comment (iziemba) [inline libfabric.vcxproj:571]: s/util_cuda_mem_monitor/cuda_mem_monitor/

File: prov/util/src/util_mem_monitor.c (hunk 11 above)
Comment (shefty) [reply prov/util/src/util_mem_monitor.c:152]: I question whether we can get a working solution without having the cache support multiple monitors, so I think it needs to be part of an initial set of changes.  That should also minimize the changes needed to the providers.

I'm okay with unified virtual addressing being required in order to use the cache.  We can document that restriction.  I think this is what works for most applications though.

Trying to support non-unified addressing requires a bit more work to make sure we can get that correct, and requires application changes.

File: prov/util/src/cuda_mem_monitor.c [hunk 13]
```diff
@@ -0,0 +1,165 @@
...
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include "ofi_mr.h"
+
+#ifdef HAVE_LIBCUDA
+
+#include "ofi_tree.h"
+#include "ofi_iov.h"
```
Comment (iziemba) [inline prov/util/src/cuda_mem_monitor.c:38]: Remove unused includes.

File: prov/util/src/util_mem_monitor.c (hunk 11 above)
Comment (iziemba) [reply prov/util/src/util_mem_monitor.c:152]: > I question whether we can get a working solution without having the cache support multiple monitors, so I think it needs to be part of an initial set of changes. That should also minimize the changes needed to the providers.

I'll add another commit to this PR for the implementation.

> I'm okay with unified virtual addressing being required in order to use the cache. We can document that restriction. I think this is what works for most applications though.

I can document this restriction.

File: Makefile.am [hunk 14]
```diff
@@ -118,6 +122,9 @@ util_fi_pingpong_LDADD = $(linkback)
 
 nodist_src_libfabric_la_SOURCES =
 src_libfabric_la_SOURCES =			\
+	include/ofi_hmem.h			\
+	include/ofi_hmem_cuda.h			\
+	include/ofi_hmem_system.h		\
```
Comment (shefty) [inline Makefile.am:128]: Please merge these into a single header file.  Each one only has 2 lines of code.

We can probably merge the .c files together as well.  We're adding nearly as many lines of license than code.

File: include/ofi_hmem_cuda.h [hunk 15]
```diff
@@ -0,0 +1,41 @@
...
+ * SOFTWARE.
+ */
+
+#ifndef _OFI_HMEM_CUDA_H_
+#define _OFI_HMEM_CUDA_H_
+
+#include <stdlib.h>
+
+int ofi_hmem_cuda_memcpy_htod(void *dev, const void *host, size_t size);
+int ofi_hmem_cuda_memcpy_dtoh(void *host, const void *dev, size_t size);
```
Comment (shefty) [inline include/ofi_hmem_cuda.h:39]: Let's reduce the amount of prefix here and just call these cuda_copy_to_dev, cuda_copy_from_dev

File: src/hmem.c [hunk 16]
```diff
@@ -0,0 +1,132 @@
...
+#include "ofi_iov.h"
+
+struct ofi_hmem_ops {
+	int (*memcpy_htod)(void *host, const void *dev, size_t size);
+	int (*memcpy_dtoh)(void *dev, const void *host, size_t size);
+};
+
+#define FI_HMEM_IFACE_TYPE_COUNT 2
+
+static struct ofi_hmem_ops hmem_ops[FI_HMEM_IFACE_TYPE_COUNT] = {
```
Comment (shefty) [inline src/hmem.c:55]: We can drop the FI_HMEM_IFACE... define and declare this as hmem_ops[] =

File: src/hmem.c [hunk 17]
```diff
@@ -0,0 +1,132 @@
...
+				       size_t size, enum fi_hmem_iface iface)
+{
+	return hmem_ops[iface].memcpy_htod(dev, host, size);
+}
+
+static inline int ofi_hmem_memcpy_dtoh(void *host, const void *dev,
+				       size_t size, enum fi_hmem_iface iface)
+{
+	return hmem_ops[iface].memcpy_dtoh(host, dev, size);
+}
```
Comment (shefty) [inline src/hmem.c:76]: These would rename to ofi_copy_from_hmem() and ofi_copy_to_hmem().

File: include/ofi_hmem.h [hunk 18]
```diff
@@ -0,0 +1,50 @@
...
+
+ssize_t ofi_copy_from_hmem_iov(void *dest, size_t size,
+			       const struct iovec *hmem_iov,
+			       enum fi_hmem_iface hmem_iface,
+			       size_t hmem_iov_count, uint64_t hmem_iov_offset);
+
+ssize_t ofi_copy_to_hmem_iov(const struct iovec *hmem_iov,
+			     enum fi_hmem_iface hmem_iface,
+			     size_t hmem_iov_count, uint64_t hmem_iov_offset,
+			     void *src, size_t size);
```
Comment (shefty) [inline include/ofi_hmem.h:48]: I think we should align on this naming (i.e. copy_to_hmem, copy_from_hmem).  See below for change details.

In the future, we can add copy_between_hmem if that's ever needed.

File: include/ofi_hmem_system.h [hunk 19]
```diff
@@ -0,0 +1,40 @@
...
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#ifndef _OFI_HMEM_SYSTEM_H_
+#define _OFI_HMEM_SYSTEM_H_
+
+#include <stdlib.h>
+
+int ofi_hmem_system_memcpy(void *dest, const void *src, size_t size);
```
Comment (shefty) [inline include/ofi_hmem_system.h:38]: Rename to ofi_hmemcpy().  Just declare as a static inline, since we're dealing with a 2 line function.

File: src/hmem.c [hunk 20]
```diff
@@ -0,0 +1,132 @@
...
+
+#include "ofi_hmem.h"
+#include "ofi_hmem_cuda.h"
+#include "ofi_hmem_system.h"
+#include "ofi.h"
+#include "ofi_iov.h"
+
+struct ofi_hmem_ops {
+	int (*memcpy_htod)(void *host, const void *dev, size_t size);
+	int (*memcpy_dtoh)(void *dev, const void *host, size_t size);
```
Comment (shefty) [inline src/hmem.c:50]: Rename to copy_to_hmem and copy_from_hmem.  If the iface is host, a device isn't involved, which makes the names a little misleading.

File: Makefile.am (hunk 14 above)
Comment (aingerson) [reply Makefile.am:128]: That's on me. I asked him to separate the c files for easier future additions since there will be more device ops to add in the future (get handle, open, handle, close handle, ipc memcpy, init, cleanup) so eventually each set of device ops will be much longer than this set.

Comment (shefty) [reply Makefile.am:128]: So, instead of 2 lines per file, we might get up to 8?  :)  I'd still combine these, especially since the definitions will be in sync.

Comment (aingerson) [reply Makefile.am:128]: lol no - specifically separating the c files

Comment (iziemba) [reply Makefile.am:128]: So merge headers but keep C files?

Comment (shefty) [reply Makefile.am:128]: At the very least kill hmem_system.c.

Comment (iziemba) [reply Makefile.am]: Done. Kept hmem_cuda.c.

**Comment by shefty** [comment]
Starting to merge what's ready through PR #5927, if discussions and reviews on the full series require more time.

File: include/ofi_mr.h [hunk 21]
```diff
@@ -111,6 +116,23 @@ struct ofi_mem_monitor {
...
 	void (*unsubscribe)(struct ofi_mem_monitor *notifier,
 			    const void *addr, size_t len);
+
+	/* Get HMEM information from the memory monitor. If a memory monitor
+	 * does not support HMEM information, -FI_ENOSYS is returned. -FI_ENOSYS
+	 * should not be treated as an error.
+	 */
+	int (*get_hmem_info)(struct ofi_mem_monitor *notifier,
+			     const void *addr, size_t len,
+			     union ofi_mr_hmem_info *hmem_info);
```
Comment (shefty) [inline include/ofi_mr.h:128]: If the monitor doesn't need the hmem info, it should just return success, so we don't have 2 non-error return values to check.  The patch does not make the purpose of this call clear.  The hmem_info is part of ofi_mr_entry, so we have it once we find the entry.  We just need to pass it into the valid callback.

I expected to see this passed into the subscribe and unsubscribe calls, so that it can be initialized as part of the subscription process.

File: include/ofi_mr.h [hunk 22]
```diff
@@ -106,6 +106,7 @@ struct ofi_mr_cache;
 
 union ofi_mr_hmem_info {
 	uint64_t reserved;
+	uint64_t cuda_id;
```
Comment (shefty) [inline include/ofi_mr.h:109]: reserved is no longer needed

File: include/ofi_mr.h [hunk 23]
```diff
@@ -177,6 +178,10 @@ void ofi_memhooks_cleanup(void);
 
 extern struct ofi_mem_monitor *memhooks_monitor;
 
+/* Cuda memory monitor. */
+extern struct ofi_mem_monitor *cuda_monitor;
+void cuda_monitor_init(void);
+void cuda_monitor_teardown(void);
```
Comment (shefty) [inline include/ofi_mr.h:185]: Please call cleanup to align with rest of calls.

File: prov/util/src/cuda_mem_monitor.c [hunk 24]
```diff
@@ -0,0 +1,162 @@
...
+#include "ofi_mr.h"
+
+#ifdef HAVE_LIBCUDA
+
+#include <cuda.h>
+#include <cuda_runtime.h>
+
+struct cuda_mm {
+	struct ofi_mem_monitor mm;
+};
```
Comment (shefty) [inline prov/util/src/cuda_mem_monitor.c:42]: There's no need to declare a struct with a single field.

File: prov/util/src/util_mem_monitor.c (hunk 11 above)
Comment (shefty) [reply prov/util/src/util_mem_monitor.c:158]: We can separate the changes to util_mem_monitor.c.  Those changes will need to have the multiple monitor support.

File: src/hmem.c [hunk 25]
```diff
@@ -126,3 +131,54 @@ ssize_t ofi_copy_to_hmem_iov(const struct iovec *hmem_iov,
...
+			"CUcontext does not support unified virtual addressining\n");
+		return -FI_EINVAL;
+
+	default:
+		FI_WARN(&core_prov, FI_LOG_CORE,
+			"Unhandle cuPointerGetAttribute return code: ret=%d\n",
+			cuda_ret);
+		return -FI_EINVAL;
+	}
+#endif /* HAVE_LIBCUDA */
```
Comment (shefty) [inline src/hmem.c:180]: Let's move the cuda code into the hmem_cuda file.  It seems like we should be able to use the cuda_valid() call for this, if we update that call to accept a NULL hmem_info parameter.

File: include/ofi_mr.h (hunk 21 above)
Comment (iziemba) [reply include/ofi_mr.h:128]: Ok. I'll remove get_hmem_info() and have ofi_mr_hmem_info initialized during subscription.

**Comment by iziemba** [comment]
> Starting to merge what's ready through PR #5927, if discussions and reviews on the full series require more time.

If you want, I can break out this PR into smaller PRs.

File: src/hmem.c (hunk 25 above)
Comment (iziemba) [reply src/hmem.c:180]: I can move this code into hmem_cuda. But, this code is different than the cuda code in cuda_mm_valid(). This function calls cuPointerGetAttribute with CU_POINTER_ATTRIBUTE_MEMORY_TYPE whereas cuda_mm_valid calls cuPointerGetAttribute with CU_POINTER_ATTRIBUTE_BUFFER_ID.

File: include/ofi_mr.h (hunk 23 above)
Comment (iziemba) [reply include/ofi_mr.h:185]: Will do.

File: prov/util/src/util_mem_monitor.c (hunk 11 above)
Comment (iziemba) [reply prov/util/src/util_mem_monitor.c:158]: I'll put these util_mem_monitor.c changes into their own commit.

File: src/hmem.c [hunk 26]
```diff
@@ -126,3 +139,29 @@ ssize_t ofi_copy_to_hmem_iov(const struct iovec *hmem_iov,
 				     hmem_iov_offset, src, size,
 				     OFI_COPY_BUF_TO_IOV);
 }
+
+int ofi_get_hmem_iface(const void *buf, enum fi_hmem_iface *iface)
+{
+	int i;
+	int ret;
+
+	/* Search backups since FI_HMEM_SYSTEM will always claim the buffer. */
```
Comment (iziemba) [inline src/hmem.c:148]: s/backups/backwards/

**Comment by iziemba** [comment]
Latest set of comments addressed. Once https://github.com/ofiwg/libfabric/pull/5927 lands, I'll remove those commits from this PR.

File: prov/util/src/util_mem_monitor.c (hunk 11 above)
Comment (shefty) [reply prov/util/src/util_mem_monitor.c]: To be clear, I don't think we want the above changes.  We need to update the cache to support multiple monitors, which I think will result in a fair number of improvements to the code.

This is just a guess at this point, but for example, I think we may end up changing ofi_mem_monitor to something like:

```
struct ofi_mem_monitor {
    blah blah existing fields... plus new valid() callback

    int (*init)(struct ofi_mem_monitor *monitor);
    void (*cleanup)(struct ofi_mem_monitor *monitor);
};
```

This would allow the cache to call init/cleanup in a loop for all active monitors.  The init() / cleanup() calls would need to be statically set.  E.g.

```
static struct ofi_uffd uffd = {
    .init = ofi_uffd_init,
    .cleanup = ofi_uffd_cleanup
};
```

The ofi_mr_cache would be updated to:

```
struct ofi_mr_cache {
    struct util_domain *domain;
    struct ofi_mem_monitor *monitor[OFI_HMEM_MAX];
    ....
};
```

The mr_cache->monitor[FI_HMEM_SYSTEM] would be set to either the uffd_monitor or memhook_monitor similar to how it is set today.  mr_cache->monitor[FI_HMEM_CUDA] would point to the only cuda option we currently have.

I haven't thought through the details, so it's possible I'm overlooking something that would make this approach impractical, such as handling a disabled monitor.

Comment (iziemba) [reply prov/util/src/util_mem_monitor.c]: This is similar to the approach I took in the _core/mr_cache: Support multiple monitors per cache_ commit. I'll look at adding in the init and cleanup ops to ofi_mem_monitor.

File: include/ofi_mr.h [hunk 27]
```diff
@@ -227,6 +236,7 @@ struct ofi_mr_entry {
 	unsigned int			subscribed:1;
 	int				use_cnt;
 	struct dlist_entry		list_entry;
+	union ofi_mr_hmem_info		hmem_info;
```
Comment (shefty) [inline include/ofi_mr.h:255]: It looks like it would make sense to merge this patch with a modified version of the previous patch that adds the iface field to mr_entry rather than mr_info.

File: include/ofi_mr.h [hunk 28]
```diff
@@ -173,6 +173,12 @@ void ofi_memhooks_cleanup(void);
 
 extern struct ofi_mem_monitor *memhooks_monitor;
 
+/* Cuda memory monitor. */
+extern struct ofi_mem_monitor *cuda_monitor;
+void cuda_monitor_setup(void);
+void cuda_monitor_teardown(void);
+int cuda_monitor_init(void);
+void cuda_monitor_cleanup(void);
```
Comment (shefty) [inline include/ofi_mr.h:182]: We should only need init / cleanup, not init + setup and cleanup + teardown.

File: include/ofi_hmem.h [hunk 29]
```diff
@@ -37,13 +37,19 @@
...
 int cuda_copy_from_dev(void *host, const void *dev, size_t size);
+int cuda_query(const void *buf);
 
 static inline int ofi_memcpy(void *dest, const void *src, size_t size)
 {
 	memcpy(dest, src, size);
 	return FI_SUCCESS;
 }
 
+static inline int ofi_query(const void *buf)
```
Comment (shefty) [inline include/ofi_hmem.h:48]: Let's give this a more descriptive name that indicates what is being queried or checked.  In this case, the function is trying to determine if the address is of a specific hmem type (ofi_is_hmem_iface? that would align with the higher level function name being used).

Also, I would change the 'buf' parameter to 'addr', since no size is given.  Though we could argue that a size should be provided, so that the entire range can be checked.

cuda_query and the query handler should align with the final proposal

File: src/hmem.c [hunk 30]
```diff
@@ -41,16 +41,24 @@
 struct ofi_hmem_ops {
 	int (*copy_to_hmem)(void *dest, const void *src, size_t size);
 	int (*copy_from_hmem)(void *dest, const void *src, size_t size);
+
+	/* Return FI_SUCCESS if the HMEM iface owns the buffer. Return
+	 * -FI_ENOSYS if the HMEM iface does not own the buffer. Everything else
```
Comment (shefty) [inline src/hmem.c:46]: ENOSYS means that the function isn't implemented.  ENODATA is a better option, though this might be okay returning a bool.  If there's an error, we should log it via FI_WARN.  In that case, we can default back to a system buffer, which is most likely.

File: src/hmem.c [hunk 31]
```diff
@@ -126,3 +139,29 @@ ssize_t ofi_copy_to_hmem_iov(const struct iovec *hmem_iov,
...
 				     OFI_COPY_BUF_TO_IOV);
 }
+
+int ofi_get_hmem_iface(const void *buf, enum fi_hmem_iface *iface)
+{
+	int i;
+	int ret;
+
+	/* Search backwards since FI_HMEM_SYSTEM will claim the buffer. */
+	for (i = ARRAY_SIZE(hmem_ops) - 1; i >= FI_HMEM_SYSTEM; i--) {
```
Comment (shefty) [inline src/hmem.c:149]: If we make the is_hmem_iface call return a bool, this function could be updated to return the fi_hmem_iface directly.  I haven't looked ahead in the patches to see what the code does if this function actually returns an error.

In any case, we don't need to iterate over FI_HMEM_SYSTEM and can just return that directly if the other hmem checks fail.

File: prov/util/src/util_mr_cache.c [hunk 32]
```diff
@@ -73,29 +73,13 @@ static int util_mr_find_overlap(struct ofi_rbmap *map, void *key, void *data)
...
-	pthread_mutex_unlock(&cache->lock);
-}
-
 /* We cannot hold the monitor lock when freeing an entry.  This call
  * will result in freeing memory, which can generate a uffd event
  * (e.g. UNMAP).  If we hold the monitor lock, the uffd thread will
  * hang trying to acquire it in order to read the event, and this thread
  * will itself be blocked until the uffd event is read.
+ *
+ * Must hold cache lock.
```
Comment (shefty) [inline prov/util/src/util_mr_cache.c:87]: Please move this patch into it's own pull request.  The locking of the cache is complex, particularly when userfaultfd is concerned.  We cannot hold a lock when allocating memory, freeing memory, or invoking a call into another library that may do this (such as register / unregister memory), which the userfaultfd thread may try to acquire.  There's a reason why separate locks were being used.

Comment (iziemba) [reply prov/util/src/util_mr_cache.c:87]: Will do.

File: prov/util/src/util_mem_monitor.c (hunk 11 above)
Comment (shefty) [reply prov/util/src/util_mem_monitor.c]: I'm going to make a change to the existing code that I think will help make the memory monitor a little clearer (because I'm confusing myself).  As an example, the gist is to do the following:

ofi_memhooks_init -> rename -> ofi_memhooks_start
ofi_memhooks_cleanup -> rename -> ofi_memhooks_stop

Create new ofi_memhooks_init call with code from ofi_monitor_init.

A similar change would be made for uffd.  I don't think we need cleanup routines, but we can add those if needed.  This should help with the confusion of having 2 init type locations for the memory monitors.

It should only take me a few hours to make this change and open a PR.

Comment (iziemba) [reply prov/util/src/util_mem_monitor.c]: Sounds good.

File: include/ofi_hmem.h (hunk 29 above)
Comment (iziemba) [reply include/ofi_hmem.h:48]: Will fix. In addition, I'll change to a bool.

File: src/hmem.c (hunk 30 above)
Comment (iziemba) [reply src/hmem.c:46]: Changing to bool.

File: src/hmem.c (hunk 31 above)
Comment (iziemba) [reply src/hmem.c:149]: Will do.

**Comment by iziemba** [comment]
Latest round of comments addressed. In addition, this PR will not build until https://github.com/ofiwg/libfabric/pull/5931 lands.

File: include/ofi_mr.h (hunk 27 above)
Comment (iziemba) [reply include/ofi_mr.h:259]: Will do.

**Comment by shefty** [comment]
Please rebase PR, or open a new PR if it's easier.

**Comment by iziemba** [comment]
I am going to break up this PR into three separate PRs: libfabric support for cuPointerGetAttribute, support for CUDA memory monitor, and support for ofi_get_hmem_iface. I think this will make the patches easier to review. All comments from this PR will have been addressed in these subsequent PRs.

**Comment by iziemba** [comment]
Subsequent PRs:
https://github.com/ofiwg/libfabric/pull/5976
https://github.com/ofiwg/libfabric/pull/5977
https://github.com/ofiwg/libfabric/pull/5978
